File: problems-extend-adapt/3201-3300/3294_unrolling-a-doubly-list-from-any-node/solution_backward.go
// unrollDoublyList runs the next pointers out to the tail without collecting
// anything; the backward sweep over Prev then gathers the whole list,
// tail first. One in-place reverse turns that tail-to-head buffer into
// the answer.
func unrollDoublyList(node *DoublyListNode) []int {
	for node != nil && node.Next != nil {
		node = node.Next
	}
	values := []int{}
	for ; node != nil; node = node.Prev {
		values = append(values, node.Val)
	}
	for left, right := 0, len(values)-1; left < right; left, right = left+1, right-1 {
		values[left], values[right] = values[right], values[left]
	}
	return values
}

File: problems/3201-3300/3294_unrolling-a-doubly-list-from-any-node/solution_backward.cpp
class Solution {
  public:
    vector<int> unrollDoublyList(DoublyListNode *node) {
        // Walk `next` to the tail without collecting anything; the backward
        // sweep over `prev` then gathers the whole list, tail first. One
        // in-place reverse turns that tail-to-head buffer into the answer.
        while (node != nullptr && node->next != nullptr) {
            node = node->next;
        }
        vector<int> values;
        for (; node != nullptr; node = node->prev) {
            values.push_back(node->val);
        }
        for (int left = 0, right = (int)values.size() - 1; left < right; ++left, --right) {
            std::swap(values[left], values[right]);
        }
        return values;
    }
};
